torch/_dynamo/variables/builder.py

```diff
@@ -1683,7 +1707,18 @@ def update_dim2constraint(dim, constraint_range, debug_name):
         # Precedence: export constraints > eager constraints
         constraint = dim2constraint.get(i)
         if constraint is None:
-            if marked_dynamic and not config.allow_ignore_mark_dynamic:
+            # NB: The "not is_subclass" hack maintains legacy behavior: constraints have
+            # historically not been enforced on the outer tensors of subclasses. Removing
+            # the check results in constraint violations for tensor subclasses that manually
+            # mark_dynamic(), possibly due to the guards introduced by the assert check
+            # against outer_size. Notable examples include NestedTensor and the test subclass
+            # DoubleSizeMaybeAddGeThreeTensor.
```

Comment:
I need help with this; the hack is the only way I could get the tests passing. It's possible that relying on the assert against `outer_size` in `torch/utils/_python_dispatch.py` to simplify symbols is the wrong way to go about things. Relevant snippet:
```python
def transform_subclass(t, callback, outer_size):
    attrs, ctx = t.__tensor_flatten__()
    transformed_tensors_dict = {}
    for attr in attrs:
        transformed_tensors_dict[attr] = callback(attr, getattr(t, attr))
    sub = type(t).__tensor_unflatten__(transformed_tensors_dict, ctx, outer_size)

    # NB: Purposefully guard here to simplify the inner / outer symbols.
    # Using sym_eq() for symbolic comparison can result in an expression that's too
    # difficult to guard on, so we use == here.
    assert sub.shape == outer_size, \
        f"Expected return value from {type(t)}__tensor_unflatten__() to have " \
        f"shape equal to outer_size={outer_size}, but got: {sub.shape}"

    return sub
```

Pretty sure this is causing constraints on subclass outer tensors to fail.

Reply:
You need to share the actual error message. Just unifying two symbols should not trigger the mark_dynamic tests, see for example:

```
import torch                            
import torch._inductor
import torch._dynamo

@torch.compile(backend="eager")
def forward(x, y):
    return x + y

x = torch.randn(4, 4)
y = torch.randn(4, 4)
torch._dynamo.mark_dynamic(x, 0)
torch._dynamo.mark_dynamic(y, 0)

forward(x, y)
```

So there's something else going on here.

Reply:
```
======================================================================
ERROR: test_wrapper_subclass_guards_on_inner_tensor (__main__.SubclassTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jbschlosser/branches/ffs/torch/testing/_internal/common_utils.py", line 2652, in wrapper
    method(*args, **kwargs)
  File "/home/jbschlosser/branches/ffs/test/dynamo/test_subclasses.py", line 671, in test_wrapper_subclass_guards_on_inner_tensor
    res = fn(x)
  File "/home/jbschlosser/branches/ffs/torch/_dynamo/eval_frame.py", line 489, in _fn
    return fn(*args, **kwargs)
  File "/home/jbschlosser/branches/ffs/torch/_dynamo/eval_frame.py", line 655, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/home/jbschlosser/branches/ffs/torch/_dynamo/convert_frame.py", line 721, in _convert_frame
    result = inner_convert(frame, cache_entry, hooks, frame_state)
  File "/home/jbschlosser/branches/ffs/torch/_dynamo/convert_frame.py", line 383, in _convert_frame_assert
    compiled_product = _compile(
  File "/home/jbschlosser/branches/ffs/torch/_dynamo/convert_frame.py", line 645, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/home/jbschlosser/branches/ffs/torch/_dynamo/utils.py", line 244, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/jbschlosser/branches/ffs/torch/_dynamo/convert_frame.py", line 625, in compile_inner
    check_fn = CheckFunctionManager(
  File "/home/jbschlosser/branches/ffs/torch/_dynamo/guards.py", line 1012, in __init__
    guard.create(builder)
  File "/home/jbschlosser/branches/ffs/torch/_guards.py", line 245, in create
    return self.create_fn(builder, self)
  File "/home/jbschlosser/branches/ffs/torch/_dynamo/guards.py", line 671, in SHAPE_ENV
    guards = output_graph.shape_env.produce_guards(
  File "/home/jbschlosser/branches/ffs/torch/fx/experimental/symbolic_shapes.py", line 2856, in produce_guards
    raise ConstraintViolationError(
torch.fx.experimental.symbolic_shapes.ConstraintViolationError: Constraints violated (L['x'].size()[0])! For more information, run with TORCH_LOGS=dynamic.
  - Not all values of RelaxedUnspecConstraint(L['x'].size()[0]) are valid because L['x'].size()[0] was inferred to be equal to 2*L['x'].inner_elem.size()[0].


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True


To execute this test, run the following from the base repo dir:
     python test/dynamo/test_subclasses.py -k test_wrapper_subclass_guards_on_inner_tensor

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

----------------------------------------------------------------------
```

Reply:
My understanding is that this happens when a dim with `RelaxedUnspecConstraint` has a size that is an expression (e.g. in this case `outer_size=inner_size * 2`). It's hitting the branch with the TODO in it here:
https://github.com/pytorch/pytorch/blob/4bb3a02d02e7a15ffe65a91bc3dd80774f5dc0ff/torch/fx/experimental/symbolic_shapes.py#L2629-L2639

Reply:
Resolved by removing the TODO branch. hopefully this is okay